pkg/ctl/install/install.go

```diff
@@ -0,0 +1,435 @@
+  labels:
+    name: %s
+  name: %s
+`
+	ns := fmt.Sprintf(nsTemplate, params.Namespace, params.Namespace)
+	manifests[namespaceFileName] = []byte(ns)
+	return manifests, nil
+}
+
+func writeFluxManifests(baseDir string, manifests map[string][]byte) error{
+	if err := os.MkdirAll(baseDir, 0700); err != nil {
+		return fmt.Errorf("cannot create Flux manifests directory (%s): %s", baseDir, err)
+	}
+	for fileName, contents := range manifests {
+		fullPath := filepath.Join(baseDir, fileName)
+		if err := ioutil.WriteFile(fullPath, contents, 0600); err != nil {
+			return fmt.Errorf("failed to write Flux manifest file %s: %s", fullPath, err)
+		}
+	}
+	return nil
+}
+
+func runShell(workDir string) error {
+	shell, err := loginshell.Shell()
+	if err != nil {
+		return fmt.Errorf("failed to obtain login shell %s: %s", shell, err)
+	}
+	shellCmd := exec.Cmd{
+		Path:   shell,
+		Dir:    workDir,
+		Stdin:  os.Stdin,
+		Stdout: os.Stdout,
+		Stderr: os.Stderr,
+	}
+	return shellCmd.Run()
+}
+
+func waitForFluxToStart(ctx context.Context, opts *installFluxOpts, restConfig *rest.Config,
+	cs *kubernetes.Clientset) (ssh.PublicKey, error) {
+	fluxSelector := metav1.LabelSelector{
+		MatchExpressions: []metav1.LabelSelectorRequirement{
+			{
+				Key:      "name",
+				Operator: metav1.LabelSelectorOpIn,
+				Values:   []string{"flux"},
+			},
+		},
+	}
+	portforwarder := portforward.PortForward{
+		Labels: fluxSelector,
+		Config: restConfig,
+		Clientset: cs,
+		DestinationPort: 3030,
+		Namespace: opts.templateParams.Namespace,
+	}
+	podDeadline := time.Now().Add(30 * time.Second)
+	for ; time.Now().Before(podDeadline); time.Sleep(1 * time.Second) {
+		err := portforwarder.Start()
+		if err == nil {
+			defer portforwarder.Stop()
+			break
+		}
+		if !strings.Contains(err.Error(), "Could not find running pod for selector") {
+			logger.Warning("Flux is not ready yet (%s), retrying ...", err)
+		}
+	}
+	if time.Now().After(podDeadline) {
+		return ssh.PublicKey{},fmt.Errorf("timed out waiting for flux's pod to be created")
+	}
+	fluxURL := fmt.Sprintf("http://127.0.0.1:%d/api/flux", portforwarder.ListenPort)
+	fluxClient := client.New(http.DefaultClient, transport.NewAPIRouter(), fluxURL, client.Token(""))
+
+	// Obtain SSH key
+	var fluxGitConfig fluxapi.GitConfig
+	gitConfigDeadline := time.Now().Add(30 * time.Second)
+	for ; time.Now().Before(gitConfigDeadline); time.Sleep(100 * time.Millisecond) {
+		repoCtx, repoCtxCancel := context.WithTimeout(ctx, opts.timeout)
+		var err error
+		fluxGitConfig, err = fluxClient.GitRepoConfig(repoCtx, false)
+		repoCtxCancel()
+		if err == nil {
+			defer portforwarder.Stop()
+			break
+		}
+		logger.Warning("Flux is not ready yet (%s), retrying ...", err)
+	}
+	if time.Now().After(podDeadline) {
+		return ssh.PublicKey{},fmt.Errorf("timed out waiting for Flux to be operative")
+	}
+	return fluxGitConfig.PublicSSHKey, nil
+}
+
+func createNamespaceSynchronously(cs *kubernetes.Clientset, namespace string) error {
```

Comment:
Why is this needed here, as well as sting-file-kubectl method above?

Reply:
Please read the comment when this function is invoked